fibreda/readme.md

```diff
@@ -0,0 +1,512 @@
+type Store interface {
+  Put(ctx context.Context, key [32]byte, value []byte, safety time.Duration) error
+  Promote(ctx context.Context, key [32]byte) (promoted bool, err error)
+  Get(ctx context.Context, key [32]byte) ([]byte, bool, error)
+  RunGC(ctx context.Context)
+  Close() error
+}
+```
+
+**GC**: every minute boundary, sweep `bu/<now>/*` and `bc/<now>/*`, deleting bucket entries, pointer, and data; run Badger value-log GC opportunistically.
+
+### 8.3 Upload verification pipeline (mandatory)
+
+1. **Dedup** by `commitment` (if already present → `deduplicated=true`, return extant attestation if desired).
+2. Build **rsema1d verification context** from `rlc_orig`.
+3. For each row:
+
+    * Verify **Merkle proof** vs `commitment`.
+    * Verify **encoding soundness** using Context‑Based Verification with the RLC context.
+    * **Why RLC check**: detects adversarial encodings that pass Merkle but are off‑codeword.
+4. Persist rows with `expire_at = now + unconfirmed_ttl + safety_buffer` (batch write + sync).
+5. Compute `service_epoch_minute`; **sign** domain‑tagged bytes; return **attestation(s)**.
+
+### 8.4 Get pathway
+
+* Accept `bitmap` or `indices`. If neither provided, return **all** stored rows for the commitment.
+* Partial fulfillment allowed; include `missing_indices`.
+
+### 8.5 Promotion Listener
+
+* Subscribe to on-chain **PFB** events.
+* On observed `commitment`, call `Promote()` to move from `bu/*` to `bc/*` and update pointer `p/*`.
+* Optionally **issue a new attestation** for the confirmed period and expose via `GetAttestation(Commitment)` or piggyback on a promotion ack.
+
+### 8.6 API limits & fairness
+
+* Enforce `server_rows_per_message_limit` and maximum serialized request size.
+* **Rate-limit** ingress to **\~10 MiB/s** (configurable). With \~**300 KiB**/req worst-case, this is **\~33 RPS**. This is **conservative** because not all requests are max size; start here and raise with telemetry.
+* Include `backoff_ms` in responses to guide clients under pressure.
+
+### 8.7 Storage roadmap
+
+* **SQL deletes are not relevant in v1**. We **start with Badger**. Later we may prototype **SQL-light** and **epoch-bucket** stores and pick based on measured throughput and GC performance.
+
+### 8.8 Operational notes
+
+* Use Badger write batches; ACK only after `Flush()+Sync`.
+* Value‑log GC after sweeps or periodically until no‑rewrite.
+* Health endpoints: `/healthz` (process up), `/readyz` (DB open, GC loop running), metrics endpoint (Prometheus).
+
+---
+
+## 9) Security considerations
+
+* **Attestation domain** prevents cross‑protocol replay; `chain_id` prevents cross‑network reuse; `service_epoch_minute` binds to a freshness window.
+* **Consensus keys** (ed25519) are the identity; we do not bind endpoint/IP in v1. If needed later, add a nonce or endpoint hash in v2.
+* **Row verification** includes Context‑Based Verification which is necessary to prevent malleability via valid‑looking but non‑decodable rows.
```

Comment:
similar to comment above I think we should use some other terminology besides "context based verification" as that is vague